### PR TITLE
#162 make rooms test cases

### LINKED ISSUE #162: rooms.v2.js 테스트 케이스 제작
# Summary 

rooms.js를 위한 테스트 케이스를 제작합니다

# Tasks 

- [x] rooms.js 테스트를 위한 테스트 케이스 제작

### PR DESCRIPTION
**Pull request by chlehdwon**
# Summary 
- rooms.js의 handler들을 테스트하는 코드를 추가하였습니다.
- test case에 대한 data들을 따로 만들어주어 DB에 영향을 주지 않도록 하였습니다.
- app 관련 test들은 구조 상 어쩔 수 없이 저렇게 진행해야 할 것 같습니다!

It closes #162

# Extra info 
make test cases for createHandler - y
make test cases for infoHandler- y
make test cases for joinHandler- y
make test cases for abortHandler- y
make test cases for searchHandler- y
make test cases for searchByUserHandler- y
make test cases for commitPaymentHandler- y
make test cases for settlementHandler- y
~~refactor codes with using exported app - n~~
refactor codes with only using test data - y

# Images or Screenshots 

# Further Work 

- make users.js test cases 
- add more detail corner cases

### CODE REVIEW

File: test/auth.replace.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const authHandlers = require("../src/service/auth.replace");
 const { userModel } = require("../src/db/mongo");
 const security = require("../security");
 
-describe("auth.replace handler", function () {
+describe("[auth.replace] 1.sparcsssoHandler", function () {
```
Comment (14KGun) [inline test/auth.replace.js:7]: mocha code에서는 화살표 함수 사용 못하나요??

참고 : https://www.notion.so/sparcs/ES6-3d8bbfc93f0c4630af19268ce5e8df77

File: test/rooms.js [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+      kaist: "20180668",
+      sparcs: "",
+      facebook: "",
+      twitter: "",
+    },
+    email: username + ".kaist.ac.kr",
+  });
+
+  return testUser;
+};
```
Comment (14KGun) [inline test/rooms.js:226]: 좋은 함수 인 것 같습니다 !!

다만, `rooms.js` 파일 외에서도 사용할 수 있도록 모듈 파일 만들어서 함수 `import` 해서 사용하면 좋을 것 같아요!

File: test/rooms.js [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+const express = require("express");
+const roomsHandlers = require("../src/service/rooms");
+const { userModel, roomModel, locationModel } = require("../src/db/mongo");
+const { generateProfileImageUrl } = require("../src/modules/modifyProfile");
+const app = express();
+
+describe("[rooms] 1.createHandler", function () {
+  it("should create room", async function () {
+    let testUser1 = userGenerator("test1");
+    await testUser1.save();
```
Comment (14KGun) [inline test/rooms.js:11]: `testUser1.save()`를 `userGenerator`에서 하는 것이 좋지 않을까요?

File: test/rooms.js [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      },
+    };
+
+    await roomsHandlers.createHandler(req, res);
+  });
+});
+
+describe("[rooms] 2.infoHandler", function () {
+  it("should return information of room", async function () {
+    let testUser1 = await userModel.findOne({ id: "test1" });
```
Comment (14KGun) [inline test/rooms.js:40]: `const`로 정의하셔도 될 것 같습니다.

참고: https://www.notion.so/sparcs/JavaScript-8f43dd503d674044b1a4420a752d7c83

**Review by 14KGun — COMMENTED** [review]
수고하셨습니다 ~~

**Comment by 14KGun** [comment]
@withSang @chlehdwon 
테스트 코드를 보니 github action과 동일한 테스트 환경을 구축하려면 매번 로컬에서 db dump 파일을 불러와야 하는데 불편함이 있을 것 같습니다.

상님 `taxiSampleGanerator` taxi-back으로 이전하고 npm으로 동일하게 작동하도록 하는거 어떠신가요??

`npm test`에서는 sampleGenerator 돌리고 테스트하도록 수정하고요!!

**Comment by withSang** [comment]
@14KGun 오 필요성에 공감합니다. 오늘밤에 다 끝내둘게요!!

File: test/auth.replace.js (hunk 1 above)
Comment (chlehdwon) [reply test/auth.replace.js:7]: 전에 하시던 분 코드 스타일 따라가다 보니 신경을 못 쓴 것 같아요! 전체적으로 ES6에 맞춰 수정해보도록 하겠습니다!

File: test/rooms.js (hunk 3 above)
Comment (chlehdwon) [reply test/rooms.js:11]: 맞는 것 같아요! 수정하겠습니다

**Comment by chlehdwon** [comment]
일단은 말씀해주신 문제 다 해결하였습니다!

**Comment by chlehdwon** [comment]
어제 rooms.js가 수정되면서 test에 오류가 생겨버렸네요... 맞춰서 고치긴 했는데 여기다가 pull 해오고 고쳐서 올려도 괜찮을까요??

**Comment by 14KGun** [comment]
> rooms.js가 수정되면서 test에 오류가 생겨버렸네요... 맞춰서 고치긴 했는데 여기다가 pull 해오고 고쳐서 올려도

너무 좋은 것 같습니다!!

그리고 [#184 이슈](https://github.com/sparcs-kaist/taxi-back/issues/184) 메인 브랜치로 병합 이후, 다시 pull 받아오시는 고 작업 이어하시는 것 어떠신가요??

**Comment by chlehdwon** [comment]
@14KGun 


> > rooms.js가 수정되면서 test에 오류가 생겨버렸네요... 맞춰서 고치긴 했는데 여기다가 pull 해오고 고쳐서 올려도
> 
> 너무 좋은 것 같습니다!!
> 
> 그리고 [#184 이슈](https://github.com/sparcs-kaist/taxi-back/issues/184) 메인 브랜치로 병합 이후, 다시 pull 받아오시는 고 작업 이어하시는 것 어떠신가요??

넵넵! 그러면 그 때 까지 보류하고 다른 test case 작업하는 걸로 할게요

File: test/rooms.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+  it("should create room which departs after 1 minute", async () => {
+    const testUser1 = await userGenerator("test1");
+    const testFrom = await locationModel.findOne({ koName: "대전역" });
+    const testTo = await locationModel.findOne({ koName: "택시승강장" });
+    const req = {
+      body: {
+        name: "test-room",
+        from: testFrom._id,
+        to: testTo._id,
+        time: Date.now() + 60 * 1000,
```
Comment (14KGun) [inline test/rooms.js:20]: 실제로는 방을 10분 단위로만 만들 수 있기는 합니다. ㅎ

File: test/rooms.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  });
+});
+
+// 8. test2 방에서 퇴장, 생성해준 data 모두 삭제
+describe("[rooms] 8.abortHandler", () => {
+  const removeTestData = async () => {
+    // drop all testData
+    await roomModel.deleteOne({ name: "test-room" });
+    await userModel.deleteOne({ id: "test1" });
+    await userModel.deleteOne({ id: "test2" });
```
Comment (14KGun) [inline test/rooms.js:191]: db를 clear 할 수 있는 함수가 `util.js` 에 있어서 다른 코드에서서도 공동으로 사용하도록 하면 좋을 것 같아요

File: test/rooms.js [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    };
+    const res = {
+      send: (data) => {
+        expect(data).to.has.property("name", "test-room");
+        expect(data).to.has.property("isOver", true);
+        expect(data).to.has.property("settlementTotal", 1);
+      },
+      status: (data) => {
+        expect(data).to.equal(200);
+      },
```
Comment (14KGun) [inline test/rooms.js:156]: 이런 부분 코드 중복이 많이 되는데, 코드 중복도 줄일 수 있는 방법이 있을까요?

File: test/rooms.js [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+      userId: testUser1.id,
+    };
+    const res = {
+      json: (data) => {
+        expect(data).to.has.property("ongoing");
+        expect(data).to.has.property("done");
+      },
+      status: (data) => {
+        expect(data).to.equal(200);
+      },
```
Comment (14KGun) [inline test/rooms.js:131]: status code가 200이면, `data`가 `ongoing`과 `done` property를 가지고 있는 것은 거의 확정이라서
property를 가지고 있는지 없는지 검사하는 것은 큰 의미가 없어 보입니다.

더 세부적으로 value로 무엇을 가지고 있는지도 검사하면 더 좋지 않을까요?

**Review by 14KGun — COMMENTED** [review]
수고하셨습니다 ~~~ 👍 👍

File: test/rooms.js (hunk 5 above)
Comment (chlehdwon) [reply test/rooms.js:20]: 다음 10분을 찾는 코드를 짜는 게 비효율적이기도 하고, 날짜가 넘어갈 시 테스트에 영향을 미쳐서 일단 직관적으로 1분으로 했습니다...! 10분 단위 맞추는 게 더 나으면 고치도록 할게요b

File: test/rooms.js (hunk 6 above)
Comment (chlehdwon) [reply test/rooms.js:191]: sampleGenerator에서만 모든 schema에 isTest 같은 항목 추가하고, util.js에서 isTest가 1인 항목만 지우도록 하면 좋을 것 같은데 그렇게 진행할까요?

File: test/rooms.js (hunk 7 above)
Comment (chlehdwon) [reply test/rooms.js:156]: 한번 생각해보도록 할게요!

File: test/rooms.js (hunk 8 above)
Comment (chlehdwon) [reply test/rooms.js:131]: 동의합니다! 이것도 수정할게요 :)

File: test/rooms.js (hunk 5 above)
Comment (14KGun) [reply test/rooms.js:20]: 아뇨아뇨. back에서 분이 10분 단위인지 검증하는 코드가 없어 의미 없을 듯 합니다.
안고치셔도 될 것 같아요 !!

File: test/rooms.js (hunk 6 above)
Comment (14KGun) [reply test/rooms.js:191]: 음... 테스트 용으로 스키마를 건드리는 것은 옳지 않아 보이는데,
테스트 어떻게 더 세부적으로 진행할 지 방학 때 더 논의해보도록 하죠...ㅋㅋㅋ

Comment (chlehdwon) [reply test/rooms.js]: 일단 global 변수를 통해서 해결해보았어요!

**Review by 14KGun — APPROVED** [review]
